File: scala/pal/src/main/scala/pal/GenerateWindowPal.scala
package pal

import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Path}
import scala.collection.immutable.VectorMap

/** Emit the complete windowed PAL source as an ordinary binary PEG. */
object GenerateWindowPal {
  final case class Options(output: Path, checkpoint: Option[Path] = None, resume: Option[Path] = None,
                           skipOptimize: Boolean = true)

  private val metadata: VectorMap[String, Any] = VectorMap(
    "kind" -> "window-pal", "k" -> GsBatchClock.DEFAULT_BATCH.k,
    "matching" -> GsBatchClock.DEFAULT_BATCH.matching, "flags" -> GsBatchClock.DEFAULT_BATCH.flags)

  def parse(args: Array[String]): Options = {
    var output: Option[Path] = None
    var checkpoint: Option[Path] = None
    var resume: Option[Path] = None
    var rewriting: Option[Boolean] = None
    var index = 0
    while (index < args.length) {
      args(index) match {
        case "--checkpoint" | "--resume" if index + 1 >= args.length =>
          throw new IllegalArgumentException(args(index) + " requires a path")
        case "--checkpoint" => checkpoint = Some(Path.of(args(index + 1))); index += 2
        case "--resume" => resume = Some(Path.of(args(index + 1))); index += 2
        case "--optimize" =>
          if (rewriting.contains(true)) { throw new IllegalArgumentException("--optimize conflicts with --skip-optimize") }
          rewriting = Some(false)
          index += 1
        case "--skip-optimize" =>
          if (rewriting.contains(false)) { throw new IllegalArgumentException("--skip-optimize conflicts with --optimize") }
          rewriting = Some(true)
          index += 1
        case option if option.startsWith("-") => throw new IllegalArgumentException("unknown option: " + option)
        case path if output.isEmpty => output = Some(Path.of(path)); index += 1
        case path => throw new IllegalArgumentException("unexpected argument: " + path)
      }
    }
    Options(output.getOrElse(throw new IllegalArgumentException("output path is required")), checkpoint, resume,
      rewriting.getOrElse(true))
  }

  private def elapsed(started: Long): Double = (System.nanoTime() - started).toDouble / 1000000000.0
  private def jsonString(value: String): String = "\"" + value.flatMap {
    case '\\' => "\\\\"
    case '"' => "\\\""
    case '\n' => "\\n"
    case '\r' => "\\r"
    case '\t' => "\\t"
    case c => c.toString
  } + "\""
  private def report(fields: (String, Any)*): Unit = {
    println(fields.map { case (key, value) =>
      val rendered = value match {
        case text: String => jsonString(text)
        case path: Path => jsonString(path.toString)
        case other => other.toString
      }
      jsonString(key) + ": " + rendered
    }.mkString("{", ", ", "}"))
  }

  def run(options: Options, sourceBuilder: () => Scaffold = () => Expr.share { ScaffoldWindowPal.build()._3 }): Unit = {
    val started = System.nanoTime()
    val machine0 = options.resume match {
      case Some(path) =>
        val (machine, savedMetadata) = ScaffoldArtifact.read(path)
        if (savedMetadata != metadata) {
          throw new IllegalArgumentException("checkpoint is not the complete default window PAL source")
        }
        machine
      case None => sourceBuilder()
    }
    report("phase" -> "built", "seconds" -> elapsed(started),
      "labels" -> machine0.labels.size, "pointers" -> machine0.pointers.size)
    options.checkpoint.foreach { path =>
      val nodes = ScaffoldArtifact.write(machine0, path, metadata)
      report("phase" -> "checkpointed", "file" -> path, "nodes" -> nodes, "seconds" -> elapsed(started))
    }
    val machine = if (options.skipOptimize) { machine0 } else {
      val (optimized, stats) = ScaffoldOptimize.optimize(machine0)
      report("phase" -> "optimized", "seconds" -> elapsed(started),
        "rounds" -> stats.rounds, "constant_labels" -> stats.constantLabels,
        "null_pointers" -> stats.nullPointers, "labels_removed" -> stats.labelsRemoved,
        "pointers_removed" -> stats.pointersRemoved)
      optimized
    }
    val writer = Files.newBufferedWriter(options.output, StandardCharsets.UTF_8)
    var count = 0
    try {
      machine.iterRules().foreach { rule => writer.write(rule); writer.newLine(); count += 1 }
    } finally {
      writer.close()
    }
    report("phase" -> "emitted", "file" -> options.output, "rules" -> count,
      "bytes" -> Files.size(options.output), "seconds" -> elapsed(started),
      "matching" -> GsBatchClock.DEFAULT_BATCH.matching, "flags" -> GsBatchClock.DEFAULT_BATCH.flags)
  }

  def main(args: Array[String]): Unit = run(parse(args))
}

File: docs/palindromes-in-peg/generate_window_pal.py
"""Emit the complete windowed PAL source as an ordinary binary PEG.

Uses only the derived DEFAULT_BATCH schedule. There is no input expansion,
word-length parameter, runtime callback, or empirical rate override.
Direct emission is the default; whole-source constant propagation is optional.
"""
import argparse
import gc
import json
from pathlib import Path
import time

from gs_batch_clock import DEFAULT_BATCH
from scaffold_window_pal import build
from scaffold_optimize import optimize
from symbolic_sca2peg import share_expressions
from scaffold_artifact import read as read_checkpoint, write as write_checkpoint


def main():
  parser = argparse.ArgumentParser(description=__doc__)
  parser.add_argument("output", type=Path)
  parser.add_argument("--checkpoint", type=Path, help="save the complete finite source before rewriting")
  parser.add_argument("--resume", type=Path, help="load a source checkpoint made by this generator")
  rewriting = parser.add_mutually_exclusive_group()
  rewriting.add_argument("--optimize", dest="skip_optimize", action="store_false",
                         help="apply optional whole-source constant propagation")
  rewriting.add_argument("--skip-optimize", dest="skip_optimize", action="store_true",
                         help="emit directly without constant propagation (default)")
  parser.set_defaults(skip_optimize=True)
  args = parser.parse_args()
  start = time.monotonic()
  collecting = gc.isenabled()
  gc.disable()
  try:
    metadata = dict(kind="window-pal", k=DEFAULT_BATCH.k,
                    matching=DEFAULT_BATCH.matching, flags=DEFAULT_BATCH.flags)
    if args.resume:
      machine, saved_metadata = read_checkpoint(args.resume)
      if saved_metadata != metadata:
        raise ValueError("checkpoint is not the complete default window PAL source")
    else:
      with share_expressions():
        circuit, controller, machine = build()
      del circuit, controller
    print(json.dumps(dict(phase="built", seconds=time.monotonic() - start,
                          labels=len(machine.labels), pointers=len(machine.pointers))), flush=True)
    # Release construction cycles once, then leave collection disabled while
    # traversing the immutable expression DAG for rewriting and emission.
    gc.collect()
    if args.checkpoint:
      nodes = write_checkpoint(machine, args.checkpoint, metadata)
      print(json.dumps(dict(phase="checkpointed", file=str(args.checkpoint), nodes=nodes,
                            seconds=time.monotonic() - start)), flush=True)
    if not args.skip_optimize:
      machine, stats = optimize(machine)
      print(json.dumps(dict(phase="optimized", seconds=time.monotonic() - start, **stats)), flush=True)
    with args.output.open("w", encoding="utf-8") as output:
      count = 0
      for rule in machine.iter_rules():
        output.write(rule + "\n")
        count += 1
    print(json.dumps(dict(phase="emitted", file=str(args.output), rules=count,
                          bytes=args.output.stat().st_size, seconds=time.monotonic() - start,
                          matching=DEFAULT_BATCH.matching, flags=DEFAULT_BATCH.flags)), flush=True)
  finally:
    if collecting: gc.enable()


if __name__ == "__main__":
  main()
